api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -601,17 +601,26 @@ public Builder alwaysNull(String sourceName) {
 
     // add a partition field with an auto-increment partition field id starting from
     // PARTITION_DATA_ID_START
+    Builder add(List<Integer> sourceIds, String name, Transform<?, ?> transform) {
+      return add(sourceIds, nextFieldId(), name, transform);
+    }
+
     Builder add(int sourceId, String name, Transform<?, ?> transform) {
-      return add(sourceId, nextFieldId(), name, transform);
+      return add(List.of(sourceId), name, transform);
     }
 
-    Builder add(int sourceId, int fieldId, String name, Transform<?, ?> transform) {
-      checkAndAddPartitionName(name, sourceId);
-      fields.add(new PartitionField(sourceId, fieldId, name, transform));
+    Builder add(List<Integer> sourceIds, int fieldId, String name, Transform<?, ?> transform) {
+      // we use the first entry in the source-ids list here
+      checkAndAddPartitionName(name, sourceIds.get(0));
```

Comment:
How does this work for multi arg partition field? 
Do we need a logic that accepts the `sourceIds` and resolve the name of multi arg transform?

Reply:
For now, I'm using the first element. I can update `checkAndAddPartitionName` to consider all elements in source ids.

Reply:
OK, let me refactore this part.